### ISSUE TITLE
Add a demo of direct path patching

### ISSUE THREAD

**Issue by neelnanda-io** [neelnanda-io/TransformerLens#111]
Direct path patching is like activation patching, but rather than patching in the output of component A, it acts on *pairs* of components A and B (in a layer after A). And we only patch in the output of A into the input of B, and all other components see the old output of A. I want to add a section to [Exploratory Analysis Demo](https://colab.research.google.com/github/neelnanda-io/Easy-Transformer/blob/main/Exploratory_Analysis_Demo.ipynb) demonstrating this for all pairs of heads.

Eg to do direct path patching on the query of head B, we'd add a hook saying `patched_B_query = original_B_query + (clean_A_output - corrupted_A_output) @ W_Q / layer_norm_scale`

For reference, an old PR to add it an early version of the library #49

**Comment by Young-Zhen** [comment]
hello, does TransformerLens support path patching now?

**Comment by callummcdougall** [comment]
I'm not able to prioritize this at the moment, however if you're intersted then I'd recommend checking out the [ARENA IOI](https://colab.research.google.com/drive/1KgrEwvCKdX-8DQ1uSiIuxwIiwzJuQ3Gw) notebook (which builds on the previously linked Exploratory Analysis Demo). Note that this notebook's path patching is slightly different to the proposed one; it takes twice as much time because it has to perform 2 sets of forward passes not 1 (but it does support a slightly wider range of path patching experiments).

**Comment by mukund1985** [comment]
PR #1396 implements this — adds `get_act_patch_direct_path()` and `get_act_patch_direct_path_all_sources()` with a validated demo on GPT-2 small / IOI task. It's merged into `dev`. Does the notebook update (Exploratory Analysis Demo) need to happen before this closes, or is the function + demo script sufficient?

**Comment by jlarson4** [comment]
@mukund1985 The notebook update was already completed by @danra, so this issue is ready to close. Thank you both for your work on this.
